File: 747-min-cost-climbing-stairs/min-cost-climbing-stairs.cpp
class Solution {
public:
    int minCostClimbingStairs(vector<int>& cost) {
        // cost[i] = min(dp[i-2] + cost[i-2], dp[i-1] + cost[i-1])
        int n = cost.size();

        int prev = 0, curr = 0;
        vector<int> dp(n+1, 0); // as it is only counted if we reach outside the array

        for (int i = 2; i <= n; i++) {
            dp[i] = min(dp[i-2] + cost[i-2], dp[i-1] + cost[i-1]);
        }

        return dp[n];
    }
};

File: 747-min-cost-climbing-stairs/min-cost-climbing-stairs.py
class Solution:
    def minCostClimbingStairs(self, cost: List[int]) -> int:
        # must reach n+1
        n = len(cost)

        if n == 2:
            return min(cost)
        
        prev, curr = cost[0], cost[1]

        for i in range(2, n):
            temp = cost[i] + min(prev, curr)
            prev = curr
            curr = temp
        
        return min(prev,curr)